src/Microsoft.SqlTools.ServiceLayer/BatchParser/BatchParserWrapper.cs

```diff
@@ -111,6 +117,73 @@ public sealed class BatchParserWrapper : IDisposable
             return batchDefinitionList;
         }
 
+        private static int GetMaxStartLine(IList<Tuple<int, int>> positions) 
+        {
+            int highest = 0;
+            foreach (var position in positions) 
+            {
+                if (position.Item1 > highest) 
+                {
+                    highest = position.Item1;
+                }
+            }
+            return highest;
+        }
+
+        /// <summary>
+        /// Gets offsets for all batches
+        /// </summary>
+        private static List<int> GetOffsets(string content, IList<Tuple<int, int>> positions) 
+        {
+
+            List<int> offsets = new List<int>();
+            int count = 0;
+            int offset = 0;
+            bool foundAllOffsets = false;
+            using (StringReader reader = new StringReader(content)) {
+                while (!foundAllOffsets)
+                 {
+                    for (int i = 0; i <= GetMaxStartLine(positions); i++)
+                    {
+                        int ch;
+                        while (true)
+                        {
+                            if (positions[count].Item1 == i) {
```

Comment:
personally I think this function is a bit too complicated to easily understand.  For example, there are 3 nested loops, one of the loops is "infinite" condition, the inner loop has multiple break statements, various counters are being incremented conditionally, and there's a lack of comments explaining what's going on.

I'd recommend to try to split this method into a few smaller methods, clean up the loops to avoid the complicated branching, add detailed comments, and add sufficient test cases to make sure this method has near-100% code coverage.